### ISSUE TITLE
autocall a number

### ISSUE THREAD

**Issue by prathibhacdac** [InnovateAsterisk/Browser-Phone#479]
How to auto call a number on opening the browser phone application? Pass the number to be called as a parameter to browser phone application.

**Comment by prathibhacdac** [comment]
Instead of user typing and dialling the other end, the call should go automatically in click of a button.

**Comment by InnovateAsterisk** [comment]
This would depend on how you have loaded the phone, if it's a full page as a tab, then you would have to write an extension in order to access the tabs array.

If you have loaded the phone as an IFRAME (loaded from the same domain), then you can simply access the IFRAME and call ```DialByLine(...)``` method on the contentDocument and pass the number to dial in there.


```
let phoneFrame = document.querySelector("iframe").contentDocument;
phoneFrame.DialByLine(...);
```

https://github.com/InnovateAsterisk/Browser-Phone/blob/12061962af0ba09899c432cdc1e18a2f2297740e/Phone/phone.js#L8710

**Comment by prathibhacdac** [comment]
How to hide the iframe and return to the main screen when the call ends?

**Comment by InnovateAsterisk** [comment]
This page contains all the events emitted in the global space.

This is the one your are looking for:
https://github.com/InnovateAsterisk/Browser-Phone/blob/12061962af0ba09899c432cdc1e18a2f2297740e/Phone/index.html#L90

So this would execute in the iframe global space. From here you would refer to the container page `parent.` and then unload the iframe. You would need to do your own checks for calls etc

**Comment by prathibhacdac** [comment]
Can you explain how to implement this. I'm not clear.

**Comment by InnovateAsterisk** [comment]
I'll add a file to the samples for a click-to-dial style setup, for the moment, this is the src:

popup.html
```
<!DOCTYPE html>
<html>
    <head>
        <title>Browser Phone (Popup)</title>
        <meta name="description" content="Browser Phone is a fully featured browser based WebRTC SIP phone for Asterisk. Designed to work with Asterisk PBX. It will connect to Asterisk PBX via web socket, and register an extension.  Calls are made between contacts, and a full call detail is saved. Audio and Video Calls can be recorded locally.">

        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>

        <meta http-equiv="Pragma" content="no-cache"/>
        <meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate"/>
        <meta http-equiv="Expires" content="0"/>

        <link rel="icon" href="favicon.ico">

        <!-- You Ouwn Scripts -->
        <link rel="stylesheet" type="text/css" href="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery-ui.min.css"/>
        <style>
            body{
                font-family: Arial, Helvetica, sans-serif;
            }
            p{
                text-align: center;
            }
            .ui-dialog .ui-dialog-content{
                padding: 0px !important;
                overflow: hidden !important;
            }
        </style>
        <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery-3.3.1.min.js"></script>
        <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery-ui.min.js"></script>
        <script lang="javascript">
                let width = 400;
                let height = 700;

                let windowObj = null;
        </script>
        <script lang="javascript">
            function CloseWindow(){
                if(windowObj != null){
                    windowObj.dialog("close");
                    windowObj = null;
                }
            }
            function OpenAsWindow(obj){
                CloseWindow();

                var iframe = $('<iframe/>');
                iframe.css("width", "100%");
                iframe.css("height", "100%");
                iframe.attr("frameborder", "0");
                if(obj && obj.val() != ""){
                    iframe.attr("src", "click-to-dial.html?d="+ obj.val());
                } else {
                    iframe.attr("src", "index.html");
                }

                // Create Window
                windowObj = $('<div/>').html(iframe).dialog({
                    autoOpen: false,
                    title: "Browser Phone",
                    modal: true,
                    width: width,
                    height: height,
                    resizable: true,
                    close: function(event, ui) {
                        $(this).dialog("destroy");
                        windowObj = null;
                    }
                });

                windowObj.dialog("open");

                var windowWidth = $(window).outerWidth();
                var windowHeight = $(window).outerHeight();
                var offsetTextHeight = windowObj.parent().outerHeight();

                windowObj.parent().css('left', windowWidth/2 - width/2 + 'px');
                windowObj.parent().css('top', windowHeight/2 - offsetTextHeight/2 + 'px');

                if(windowWidth <= width) {
                    windowObj.parent().css('left', '0px');
                } 
                if(windowHeight <= offsetTextHeight) {
                    windowObj.parent().css('top', '0px');
                }

            }
        </script>
    </head>

    <body>
        <div id=content>
            <p>This is a sample of how the phone can be loaded into you own web site.</p>
            <p>
              <button onclick="OpenAsWindow()">Open As Window</button>
            </p>

            <p><input type="tel" id="dialNumber"><button onclick="OpenAsWindow($('#dialNumber'))">Click to Dial</button></p>
        </div>
    </body>
</html>
```


click-to-dial.html
```
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">

        <title>Browser Phone</title>
        <meta name="description" content="Browser Phone is a fully featured browser based WebRTC SIP phone for Asterisk. Designed to work with Asterisk PBX. It will connect to Asterisk PBX via web socket, and register an extension.  Calls are made between contacts, and a full call detail is saved. Audio and Video Calls can be recorded locally.">

        <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0, user-scalable=no"/>

        <!-- Cache -->
        <meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate"/>
        <meta http-equiv="Expires" content="0"/>

        <link rel="icon" type="image/x-icon" href="favicon.ico">

        <!-- Styles -->
        <link rel="stylesheet" type="text/css" href="https://dtd6jl0d42sve.cloudfront.net/lib/Normalize/normalize-v8.0.1.css"/>
        <link rel="stylesheet preload prefetch" type="text/css" as="style" href="https://dtd6jl0d42sve.cloudfront.net/lib/fonts/font_roboto/roboto.css"/>
        <link rel="stylesheet preload prefetch" type="text/css" as="style" href="https://dtd6jl0d42sve.cloudfront.net/lib/fonts/font_awesome/css/font-awesome.min.css"/>
        <link rel="stylesheet" type="text/css" href="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery-ui-1.13.2.min.css"/>
        <link rel="stylesheet" type="text/css" href="https://dtd6jl0d42sve.cloudfront.net/lib/Croppie/Croppie-2.6.4/croppie.css"/>
        <link rel="stylesheet" type="text/css" href="phone.css"/>

        <!-- Provisioning -->
        <script type="text/javascript">
            // Provision runtime options can go here.
            var phoneOptions = {
                loadAlternateLang: true,
                VoiceMailSubscribe: false,
                EnableTextMessaging: false,
                DisableFreeDial: true,
                DisableBuddies: true,
                ChatEngine: "SIMPLE",
            }

            var web_hook_on_register = function(ua){
                let urlParams = new URLSearchParams(window.location.search);
                if(urlParams.has("d")){
                    window.setTimeout(function(){
                        console.log("Performing Auto Dial: ", urlParams.get("d"));
                        /**
                         * Primary method for making a call. 
                         * @param {string} type (required) Either "audio" or "video". Will setup UI according to this type.
                         * @param {Buddy} buddy (optional) The buddy to dial if provided.
                         * @param {sting} numToDial (required) The number to dial.
                         * @param {string} CallerID (optional) If no buddy provided, one is generated automatically using this callerID and the numToDial
                         * @param {Array<string>} extraHeaders = (optional) Array of headers to include in the INVITE eg: ["foo: bar"] (Note the space after the :)
                         */
                        DialByLine("audio", null, urlParams.get("d"));
                    }, 1000);
                }
            }
            var web_hook_on_registrationFailed = function(e){
                window.parent.CloseWindow();
            }
            var web_hook_on_unregistered = function(){
                window.parent.CloseWindow();
            }
            var web_hook_on_terminate = function(session){
                window.setTimeout(function(){
                    window.parent.CloseWindow();
                }, 1000);
            }

        </script>
    </head>

    <body>
        <!-- Loading Animation -->
        <div class=loading>
            <span class="fa fa-circle-o-notch fa-spin"></span>
        </div>

        <!-- The Phone -->
        <div id=Phone></div>
    </body>

    <!-- Loadable Scripts -->
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery-3.6.1.min.js"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery-ui-1.13.2.min.js"></script>
    <script type="text/javascript" src="phone.js"></script>

    <!-- Deferred Scripts -->
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/jquery/jquery.md5-min.js" defer="true"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/Chart/Chart.bundle-2.7.2.min.js" defer="true"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/SipJS/sip-0.20.0.min.js" defer="true"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/FabricJS/fabric-2.4.6.min.js" defer="true"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/Moment/moment-with-locales-2.24.0.min.js" defer="true"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/Croppie/Croppie-2.6.4/croppie.min.js" defer="true"></script>
    <script type="text/javascript" src="https://dtd6jl0d42sve.cloudfront.net/lib/XMPP/strophe-1.4.1.umd.min.js" defer="true"></script>

</html>
```

**Comment by InnovateAsterisk** [comment]
This is now in the source code:
https://github.com/InnovateAsterisk/Browser-Phone/blob/master/Phone/index.html

Hosted:
https://www.innovateasterisk.com/phone/popup.html

**Comment by prathibhacdac** [comment]
Is it possible to remove the iframe when the call ends?

**Comment by InnovateAsterisk** [comment]
If you use the click-to-dial button, it does.

**Comment by prathibhacdac** [comment]
Thank you.

**Comment by prathibhacdac** [comment]
How to handle the incoming call if the browser phone application is not opened?

**Comment by InnovateAsterisk** [comment]
Then you need to show/hide the window, rather than open/close. I don't have the code at hand, but the library in use there is jQuery UI. 

https://jqueryui.com/dialog/

Also make sure that you change the action of the close button to "hide"

**Comment by prathibhacdac** [comment]
Is it possible to close the parent window containing the Click to Dial button?

**Comment by prathibhacdac** [comment]
> Then you need to show/hide the window, rather than open/close. I don't have the code at hand, but the library in use there is jQuery UI.
> 
> https://jqueryui.com/dialog/
> 
> Also make sure that you change the action of the close button to "hide"

Show we keep the browser always on alive but in hidden mode?

**Comment by prathibhacdac** [comment]
> > Then you need to show/hide the window, rather than open/close. I don't have the code at hand, but the library in use there is jQuery UI.
> > https://jqueryui.com/dialog/
> > Also make sure that you change the action of the close button to "hide"
> 


Where should we implement the jQueryUI dialog?

**Comment by InnovateAsterisk** [comment]
If you want to keep the phone connected, and hidden in the background, you can add an open function (like above), and then directly after that include a hide function, on the page load. This will log the phone in, and leave it registered for calls. Then use the `web_hook_on_invite` to show the phone for an inbound call. (It will ring and show the answer option.)

https://github.com/InnovateAsterisk/Browser-Phone/blob/12061962af0ba09899c432cdc1e18a2f2297740e/Phone/index.html#L78

PS. Just remember that if you put the SIP credentials in an open public space, anyone can get them. Make sure you are expecting people to be logged in. Something like a CRM or even better a private CRM would be a good place. Then also remember to provision specific user details per user, so you can cycle credentials frequently.

**Comment by prathibhacdac** [comment]
Calling the HideWindow() function on page load. When should we call the OpenAsWindow().

**Comment by prathibhacdac** [comment]
$(this).dialog("hide") and $(this).dialog("show") are not working.

**Comment by prathibhacdac** [comment]
Then use the `web_hook_on_invite` to show the phone for an inbound call. (It will ring and show the answer option.)

Registration done and the window gets hidden. But unable to show the phone on inbound call. Can you pls help.

**Comment by prathibhacdac** [comment]
`$(this).dialog("show") and $(this).dialog("hide") is not working`

**Comment by InnovateAsterisk** [comment]
Not sure if I understand the issue? 
if `$(this).dialog("show")` works, that’s all you need.

**Comment by prathibhacdac** [comment]
> Not sure if I understand the issue? if `$(this).dialog("show")` works, that’s all you need.
> 
> You will have to call this code from the web hook inside the phone iframe, and refer to the `parent. `Method to show the phone

Show and hide are not working

**Comment by prathibhacdac** [comment]
In web_hook_on_invite(), I'm calling window.parent.OpenAsWindow();. But when I make a call, the dialer doesn't popup.

**Comment by prathibhacdac** [comment]
Its working. Thank you for the support.

**Comment by InnovateAsterisk** [comment]
If you asking about Web Push notifications? then yes, you can register the browser for push notifications, and get Asterisk to initiate a push.

This is a page that you can use to play around with web push.
https://web-push-codelab.glitch.me/

It must be said tho, this is quite complicated stuff, and getting Asterisk to play along with this may not be so easy. You will have to register the browser for push, capture the token (and save it with the Asteris details), and then get asterisk to send a push, so that the correct browser window gets the notification. Also, when you get the notification, what are you going to do? If you intend to show the call answer window, it may be too late for Asterisk, it may have already sent the call. You may have to put Asterisk in a wait loop.

I have managed to fully achieve this in Siperb, but it take quite a lot of work, and only OpenSIPS can do what I needed. I needed it to take the inbound call, send the push, and then wait for the device(s) to come online. Only then do i send the call to the browser. If you have multiple devices (some registered and some not), then you have to handle each one in parallel and separately check each fork, to know if someone answered on any of their registrations.

**Comment by prathibhacdac** [comment]
How to prevent reregistration if its already registered and hidden? Now on passing the number to dial, it is asking for the registration details But I want to use the already registered and hidden window.

**Comment by prathibhacdac** [comment]
How to send a message to the android webview once the iframe is closed? This is to close the webview.

**Comment by InnovateAsterisk** [comment]
> How to prevent reregistration if its already registered and hidden?

If the phone is registered, you don't need to do anything, the phone will be ready to receive the call.

> Now on passing the number to dial, it is reregistering. But I want to use the already registered and hidden window.

You can check to see if the phone is visible and registered. `userAgent.isRegistered()` returns bool.

**Comment by prathibhacdac** [comment]
I am using this in Call Dispatcher module. In the Call Dispatcher, the emergency calls will be send to different vehicles. The video calling facility is provided to each vehicle assigned to a case. Whenenver the video call button is presssed, the communication with the server is established and the user registration happens and then only the call goes to the vehicle. 

But in whatsapp there is no registration on making each call. Directly call connecting and ringing appears. Is it possible to establish the same behaviour in Browser Phone.
